src/Uno.UWP/Buffers/LinearArrayPool.cs

```diff
@@ -0,0 +1,224 @@
+		private static readonly bool _canUseMemoryManager;
+		private static readonly int _defaultAutomaticBucketSize;
+		private static readonly int _defaultBucketSize;
+
+		private static IArrayPoolPlatformProvider _platformProvider = new DefaultArrayPoolPlatformProvider();
+
+		static LinearArrayPool()
+		{
+			_automaticMemoryManagement = WinRTFeatureConfiguration.ArrayPool.EnableAutomaticMemoryManagement;
+
+			_canUseMemoryManager = _automaticMemoryManagement && _platformProvider.CanUseMemoryManager;
+
+			_defaultAutomaticBucketSize = WinRTFeatureConfiguration.ArrayPool.DefaultAutomaticMaxNumberOfArraysPerBucket;
+			_defaultBucketSize = WinRTFeatureConfiguration.ArrayPool.DefaultMaxNumberOfArraysPerBucket;
+		}
+
+		public LinearArrayPool(int baseArraySize, int bucketCount, int bucketSize, bool enableTrimming)
+		{
+			if (baseArraySize <= 0 || bucketCount <= 0 || bucketSize <= 0)
+			{
+				throw new ArgumentException("baseArraySize, bucketCount or bucketSize must be greater than 0.");
+			}
+
+			_baseArraySize = baseArraySize;
+
+			_buckets = new Bucket[bucketCount];
+
+			for (var x = 0; x < bucketCount; x++)
+			{
+				_buckets[x] = new Bucket(bucketSize);
+			}
+
+			if (enableTrimming)
+			{
+				_requestTrimCallbackRegistration = 1;
+			}
+		}
+
+		public static LinearArrayPool<T> CreateAutomaticallyManaged(int baseArraySize, int bucketCount)
+			=> new LinearArrayPool<T>(
+				baseArraySize,
+				bucketCount,
+				_canUseMemoryManager ? _defaultAutomaticBucketSize : _defaultBucketSize,
+				_canUseMemoryManager);
+
+		public T[] Rent(int multipleOfBaseSize)
+		{
+			Debug.Assert(multipleOfBaseSize >= 0 && (multipleOfBaseSize % _baseArraySize) == 0);
+
+			if (multipleOfBaseSize != 0)
+			{
+				var index = (multipleOfBaseSize / _baseArraySize) - 1;
+
+				if (index < _buckets.Length)
+				{
+					return _buckets[index].TryPop() ?? new T[multipleOfBaseSize];
+				}
+
+				return new T[multipleOfBaseSize];
+			}
+
+			return Array.Empty<T>();
+		}
+
+		public void Return(T[] array, bool clearArray)
+		{
+			if (array != null)
+			{
+				if (array.Length != 0)
+				{
+					Debug.Assert(array.Length % _baseArraySize == 0);
+
+					var index = (array.Length / _baseArraySize) - 1;
+
+					if (index < _buckets.Length)
+					{
+						if (clearArray)
+						{
+							Array.Clear(array);
+						}
+
+						_buckets[index].TryPush(array);
+
+						if (Interlocked.Exchange(ref _requestTrimCallbackRegistration, 0) == 1)
+						{
+							_platformProvider.RegisterTrimCallback(pool => ((LinearArrayPool<T>)pool).Trim(), this);
+						}
+					}
+				}
+			}
+			else
+			{
+				throw new ArgumentNullException(nameof(array));
+			}
+		}
+
+		private bool Trim()
+		{
+			var memoryUsage = _platformProvider.AppMemoryUsageLevel;
+
+			var now = _platformProvider.Now;
+
+			for (var x = 0; x < _buckets.Length; x++)
+			{
+				_buckets[x].Trim(memoryUsage, now);
+			}
+
+			return true;
+		}
+
+		private class Bucket
```

Comment:
Consider sealing this class. It should result in a more efficient code. The difference might not be large compared to other parts of the code, it's a simple change anyway